File: samples/python/24.bot-authentication-msgraph/dialogs/logout_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs import DialogTurnResult, ComponentDialog, DialogContext
from botbuilder.core import BotFrameworkAdapter
from botbuilder.schema import ActivityTypes


class LogoutDialog(ComponentDialog):
    def __init__(self, dialog_id: str, connection_name: str):
        super(LogoutDialog, self).__init__(dialog_id)

        self.connection_name = connection_name

    async def on_begin_dialog(
        self, inner_dc: DialogContext, options: object
    ) -> DialogTurnResult:
        result = await self._interrupt(inner_dc)
        if result:
            return result
        return await super().on_begin_dialog(inner_dc, options)

    async def on_continue_dialog(self, inner_dc: DialogContext) -> DialogTurnResult:
        result = await self._interrupt(inner_dc)
        if result:
            return result
        return await super().on_continue_dialog(inner_dc)

    async def _interrupt(self, inner_dc: DialogContext):
        if inner_dc.context.activity.type == ActivityTypes.message:
            text = inner_dc.context.activity.text.lower()
            if text == "logout":
                bot_adapter: BotFrameworkAdapter = inner_dc.context.adapter
                await bot_adapter.sign_out_user(inner_dc.context, self.connection_name)
                await inner_dc.context.send_activity("You have been signed out.")
                return await inner_dc.cancel_all_dialogs()

File: archive/samples/java_springboot/46.teams-auth/src/main/java/com/microsoft/bot/sample/teamsauth/LogoutDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.teamsauth;

import java.util.concurrent.CompletableFuture;

import com.microsoft.bot.builder.BotFrameworkAdapter;
import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.dialogs.ComponentDialog;
import com.microsoft.bot.dialogs.DialogContext;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.schema.ActivityTypes;

public class LogoutDialog extends ComponentDialog {

    private final String connectionName;

    public LogoutDialog(String id, String connectionName) {
        super(id);
        this.connectionName = connectionName;
    }


    @Override
    protected CompletableFuture<DialogTurnResult> onBeginDialog(DialogContext innerDc, Object options) {
        return interrupt(innerDc).thenCompose(result -> {
            if (result != null) {
                return CompletableFuture.completedFuture(result);
            }

            return super.onBeginDialog(innerDc, options);
        });
    }

    @Override
    protected CompletableFuture<DialogTurnResult> onContinueDialog(DialogContext innerDc) {
        return interrupt(innerDc).thenCompose(result -> {
            if (result != null) {
                return CompletableFuture.completedFuture(result);
            }

            return super.onContinueDialog(innerDc);
        });
    }

    private CompletableFuture<DialogTurnResult> interrupt(DialogContext innerDc) {
        if (innerDc.getContext().getActivity().getType().equals(ActivityTypes.MESSAGE)) {
            String text = innerDc.getContext().getActivity().getText().toLowerCase();

            // Allow logout anywhere in the command
            if (text.contains("logout")) {
                // The bot adapter encapsulates the authentication processes.
                BotFrameworkAdapter botAdapter = (BotFrameworkAdapter) innerDc.getContext().getAdapter();
                botAdapter.signOutUser(
                        innerDc.getContext(),
                        getConnectionName(),
                        innerDc.getContext().getActivity().getFrom().getId())
                        .thenApply(result -> innerDc.getContext().sendActivity(MessageFactory.text("You have been signed out."))
                                .thenApply(sendResult -> null));
                return innerDc.cancelAllDialogs();
            }
        }

        return CompletableFuture.completedFuture(null);
    }

    /**
     * @return the ConnectionName value as a String.
     */
    protected String getConnectionName() {
        return this.connectionName;
    }

}

